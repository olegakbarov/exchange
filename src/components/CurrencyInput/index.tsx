import * as React from "react";
import { ChangeEvent } from "react";
import styled from "styled-components/macro";
import NumberFormat from "react-number-format";

interface IProps {
  value: number;
  handleChange: (x: number) => void;
  maxValue: number;
  label?: string;
  disabled?: boolean;
}

interface IValues {
  formattedValue: string;
  value: number;
  floatValue: number;
}

const CurrencyInput = React.memo(
  ({ value, label, handleChange, maxValue, disabled }: IProps) => {
    const onChange = React.useCallback(
      (e: ChangeEvent<HTMLInputElement>) => {
        const nextVal = Number(e.target.value);
        handleChange(nextVal);
      },
      [handleChange]
    );

    const isAllowed = (values: IValues) => {
      return values.value <= maxValue;
    };

    return (
      <Label>
        {label && label}
        <StyledInput
          disabled={disabled}
          allowNegative={false}
          decimalScale={2}
          allowLeadingZeros={false}
          value={value}
          onChange={onChange}
          isAllowed={isAllowed}
        />
      </Label>
    );
  }
);

const StyledInput = styled(NumberFormat)`
  font-size: 22px;
  background-color: ${p => p.theme.inputBgColor};
  color: ${p => p.theme.fgColor};
  padding-top: 10px;
  padding-bottom: 10px;
  padding-right: 5px;
  border-radius: 4px;
  text-align: right;
  border: none;
  width: 100%;
  margin: 7px 0;
`;

const Label = styled.label`
  display: block;
  font-size: 12px;
  text-transform: uppercase;
`;

export default CurrencyInput;
