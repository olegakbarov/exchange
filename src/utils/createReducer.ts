import { Action } from "redux";
import { Reducer } from "../types/redux";
// import { AsyncActionStatus } from '../types/asyncActions';

type RootReducer<S, A extends Action = Action> = {
  [type: string]: Reducer<S, A>;
};

// type AsyncAction = Action & {
//   status: AsyncActionStatus;
// };

// function isAsyncAction(action: Action): action is AsyncAction {
//   return action.hasOwnProperty('status');
// }

export default function createReducer<S extends {}, A extends Action = Action>(
  reducerName: string,
  reducers: RootReducer<S, A>,
  state: S,
  action: A
): S {
  Object.keys(reducers).forEach(
    a => !a && console.error("Undefined reducer found!")
  );

  let restoredState = state;

  const reducerOrMap = reducers[action.type];
  let reducer: Reducer<S, A> | undefined;

  if (reducerOrMap && typeof reducerOrMap !== "function") {
    // if (isAsyncAction(action)) {
    //   reducer = reducerOrMap[action.status];
    // }
  } else {
    reducer = reducerOrMap;
  }

  const newState = reducer
    ? reducer(restoredState as S, action)
    : (restoredState as S);

  return newState;
}
