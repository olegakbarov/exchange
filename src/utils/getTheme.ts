import { Theme } from "../types/theme";
import { Themes } from "../types/enums";

export default function getTheme(name: Themes): Theme {
  switch (name) {
    case Themes.Light:
      return {
        white: "#fff",
        successGreen: "#56d256",
        errorRed: "#ff2828",
        bgColor: "#f8f8fa",
        fgColor: "#000",
        buttonBgColor: "rgb(38, 132, 255)",
        borderColor: "#eaeaea",
        inputBgColor: "#fff;"
      };

    case Themes.Dark:
      return {
        white: "#fff",
        successGreen: "#56d256",
        errorRed: "#ff2828",
        bgColor: "#000",
        fgColor: "#fff",
        buttonBgColor: "charcoal",
        borderColor: "#fff",
        inputBgColor: "#4d4d4d;"
      };
  }
}
