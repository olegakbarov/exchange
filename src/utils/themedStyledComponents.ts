import * as styledComponents from "../../node_modules/styled-components";
import { ThemedStyledComponentsModule } from "../../node_modules/@types/styled-components/index.d";

import { Theme } from "../types/theme";

const {
  default: styled,
  css,
  createGlobalStyle,
  keyframes,
  withTheme,
  ThemeProvider
} = styledComponents as ThemedStyledComponentsModule<Theme>;

export { css, createGlobalStyle, keyframes, ThemeProvider, withTheme };

export * from "../../node_modules/@types/styled-components/index.d";

export default styled;
