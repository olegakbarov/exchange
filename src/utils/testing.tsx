import * as React from "react";
import { ThemeProvider } from "styled-components/macro";
import getTheme from "./getTheme";
import { shallow, render } from "enzyme";
import { Themes } from "../types/enums";

export const shallowWithTheme = (children: any) =>
  shallow(
    <ThemeProvider theme={getTheme(Themes.Light)}>{children}</ThemeProvider>
  );

export const renderWithTheme = (children: any) =>
  render(
    <ThemeProvider theme={getTheme(Themes.Light)}>{children}</ThemeProvider>
  );
