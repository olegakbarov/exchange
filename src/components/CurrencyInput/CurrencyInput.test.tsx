import * as React from "react";
import "jest-styled-components";
import CurrencyInput from "./index";
import { shallowWithTheme, renderWithTheme } from "../../utils/testing";

describe("CurrencyInput", () => {
  it("Shallow render snapshots are equal", () => {
    const tree = shallowWithTheme(
      <CurrencyInput
        disabled={false}
        value={39.01}
        handleChange={() => {}}
        maxValue={100}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it("Render snapshots are equal", () => {
    const tree = renderWithTheme(
      <CurrencyInput
        disabled={false}
        value={39.01}
        handleChange={() => {}}
        maxValue={100}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
