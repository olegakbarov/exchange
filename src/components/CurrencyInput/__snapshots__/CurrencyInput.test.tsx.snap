// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput Render snapshots are equal 1`] = `
.c0 {
  font-size: 17px;
  background-color: #fff;
  color: #000;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-right: 5px;
  border-radius: 4px;
  text-align: right;
  border: none;
  width: 100%;
  margin: 7px 0;
}

<input
  class="c0"
  type="text"
  value="39.01"
/>
`;

exports[`CurrencyInput Shallow render snapshots are equal 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
